### PR TITLE
LRU refactoring

### PR DESCRIPTION
**Pull request by arutk**
rearanging lru implementation for easier journaling

Signed-off-by: Adam Rutkowski <adam.j.rutkowski@intel.com>

### CODE REVIEW

File: src/ocf_lru.c [hunk 1]
```diff
@@ -298,7 +265,11 @@ static void ocf_lru_repart_locked(ocf_cache_t cache, ocf_cache_line_t cline,
 	src_list = ocf_lru_get_list(src_part, lru_list, clean);
 	dst_list = ocf_lru_get_list(dst_part, lru_list, clean);
 
-	ocf_lru_move(cache, cline, src_part, src_list, dst_part, dst_list);
+	ocf_lru_move(cache, cline, src_list, dst_list);
+	ocf_metadata_set_partition_id(cache, cline, dst_part->id);
+	env_atomic_dec(&src_part->runtime->curr_size);
+	env_atomic_inc(&src_part->runtime->curr_size);
```
Comment (robertbaldyga) [inline src/ocf_lru.c:271]: `dst_part`

**Review by robertbaldyga — CHANGES_REQUESTED** [review]

File: src/ocf_lru.c [hunk 2]
```diff
@@ -16,11 +16,70 @@
...
+		struct ocf_lru_list *list, int change)
+{
+	ocf_cache_line_t last_hot_new, last_hot_old;
+
+	last_hot_old = list->last_hot;
+
+	if (change > 0) {
+		ENV_BUG_ON(change != 1);
+
+		if (list->last_hot == end_marker) {
```
Comment (robertbaldyga) [inline src/ocf_lru.c:31]: `unlikely()` may be a good idea in this case

**Review by robertbaldyga — APPROVED** [review]

**Review by robertbaldyga — APPROVED** [review]

File: src/ocf_lru.c [hunk 3]
```diff
@@ -810,19 +796,18 @@ void ocf_lru_hot_cline(ocf_cache_t cache, ocf_cache_line_t cline)
...
 	OCF_METADATA_LRU_WR_LOCK(cline);
 
-	if (node->next != end_marker ||
-			node->prev != end_marker ||
-			list->head == cline || list->tail == cline) {
-		remove_lru_list(cache, list, cline);
-	}
+	/* cacheline must be on the list when set_hot gets called */
+	ENV_BUG_ON(node->next == end_marker && list->tail != cline);
+	ENV_BUG_ON(node->next == end_marker && list->tail != cline);
```
Comment (mmkayPL) [inline src/ocf_lru.c:808]: Was this supposed to be a different bugon?

Comment (arutk) [reply src/ocf_lru.c]: good catch - fixed

**Review by mmkayPL — APPROVED** [review]

**Review by robertbaldyga — APPROVED** [review]

**Pull request merged**